docs/design/metrics-best-practices.md

```diff
@@ -0,0 +1,70 @@
+# Kube-State-Metrics - Timeseries best practices
+
+---
+
+Author: Manuel Rüger (<manuel@rueg.eu>)
+
+Date: October 17th 2024
+
+---
+
+## Introduction
+
+Kube-State-Metrics' goal is to provide insights into the state of Kubernetes objects by exposing them as metrics.
+This document provides guidelines with the goal to create a good user experience when using these metrics.
+
+Please be aware that this document is introduced in a later stage of the project and there might be metrics that do not follow these best practices.
+Feel encouraged to report these metrics and provide a pull request to improve them.
+
+## General best practices
+
+We follow [Prometheus](https://prometheus.io/docs/practices/naming/) best practices in terms of naming and labeloing.
+
+## Best practices for kube-state-metrics
+
+### Avoid pre-computation
+
+kube-state-metrics should expose metrics on an individual object level and avoid any sort of pre-computation unless it is required due to for example high cardinality on objects.
+By exposing raw metrics instead of counters, kube-state-metrics allows the user to have full control on how they want to use the metrics.
```

Comment:
Raw counters are usually what we want, maybe you meant "derived metrics"?

```suggestion
By exposing raw metrics instead of derived rate metrics, kube-state-metrics allows the user to have full control on how they want to use the metrics.
```

Reply:
I probably should rewrite this section a bit.
What I'm trying to describe here is that we do not want `_total` counters e.g. a `kube_pod_total` metric that counts all pods in the cluster.
Rather we prefer to expose a `kube_pod_info` metric to allow the user either run a sum on it, or even sum it by a specific attribute (e.g. per node)

Reply:
Oh, yes, that makes sense.

I think I've described this as "We prefer to not add metrics that can be derived from existing raw metrics. For example `kube_total_pods` can be computed with `count(kube_pod_info)`."